### PR TITLE
Better CC and SFZ support

### PR DESCRIPTION
**Pull request by MyBlackMIDIScore**
Changes:
- Support for CC volume, pan, damper
- Support for SFZs with multiple keys per sample
- Preparation for cutoff support (it is not implemented and has commented code that doesn't work)

### CODE REVIEW

File: core/src/voice/sampler.rs [hunk 1]
```diff
@@ -148,33 +148,43 @@ impl<S: Simd, Sampler: BufferSampler> SIMDSampleGrabber<S> for SIMDSampleGrabber
...
     Pitch: SIMDVoiceGenerator<S, SIMDSampleMono<S>>,
+    Cutoff: SIMDVoiceGenerator<S, SIMDSampleMono<S>>,
     Grabber: SIMDSampleGrabber<S>,
 {
     grabber_left: Grabber,
     grabber_right: Grabber,
 
     pitch_gen: Pitch,
 
+    cutoff: Cutoff,
```
Comment (arduano) [inline core/src/voice/sampler.rs:163]: It's probably ok to remove this cutoff code, if making a cutoff filter this wouldn't exactly be the way to do it. I can show you how to ideally approach it later. I'm pretty sure this is unused anyway so can remove

File: realtime/src/config.rs [hunk 2]
```diff
@@ -7,9 +7,9 @@ pub struct XSynthRealtimeConfig {
 impl Default for XSynthRealtimeConfig {
     fn default() -> Self {
         Self {
-            render_window_ms: 10.0,
+            render_window_ms: 20.0,
             channel_count: 16,
-            use_threadpool: false,
```
Comment (arduano) [inline realtime/src/config.rs:12]: Probably undo these changes, we can discuss better defaults if you think these aren't good, otherwise you can override them when running the examples manually

File: soundfonts/src/sfz/mod.rs [hunk 3]
```diff
@@ -54,11 +54,14 @@ pub struct RegionParamsBuilder {
...
     hivel: u8,
     key: Option<u8>,
+    lokey: Option<u8>,
+    hikey: Option<u8>,
     pitch_keycenter: Option<u8>,
     pan: i8,
     sample: Option<String>,
     default_path: Option<String>,
     loop_mode: SfzLoopMode,
+    cutoff: f32,
```
Comment (arduano) [inline soundfonts/src/sfz/mod.rs:64]: If we're making cutoff optional in the synth (e.g. different code is used for voices with and without cutoff), it might be a good idea to make this `Option<f32>`

File: core/src/soundfont/mod.rs [hunk 4]
```diff
@@ -198,25 +204,72 @@ impl SampleSoundfont {
...
-            let spawner_params = Arc::new(SampleVoiceSpawnerParams {
-                envelope: envelope_params,
-                speed_mult,
-                sample: samples[&params].clone(),
-            });
-
-            for vel in region.lovel..=region.hivel {
-                let index = key_vel_to_index(region.key, vel);
-                spawner_params_list[index] = Some(spawner_params.clone());
+            if region.lokey.is_none() && region.hikey.is_none() {
```
Comment (arduano) [inline core/src/soundfont/mod.rs:207]: As I suggested in DMs, it might be easier to just use ranges in region params rather than hi/lo values, so these if statements aren't necessary.

**Review by arduano — CHANGES_REQUESTED** [review]

File: core/src/channel/key.rs [hunk 5]
```diff
@@ -9,7 +9,7 @@ use super::{
 
 pub struct KeyData {
     key: u8,
-    voices: VoiceBuffer,
+    pub voices: VoiceBuffer,
```
[the remark this answers is not in the document]
Comment (MyBlackMIDIScore) [reply core/src/channel/key.rs:12]: Oh true. I'll change that.

File: core/src/voice/sampler.rs (hunk 1 above)
Comment (MyBlackMIDIScore) [reply core/src/voice/sampler.rs:163]: Yeah it is unused currently. I'll remove it.

File: realtime/src/config.rs (hunk 2 above)
Comment (MyBlackMIDIScore) [reply realtime/src/config.rs:12]: Oh this was staged by accident. It was part of some temporary changes I made.
I'll revert it.

File: soundfonts/src/sfz/mod.rs (hunk 3 above)
Comment (MyBlackMIDIScore) [reply soundfonts/src/sfz/mod.rs:64]: That would require if statements on each voice to check if there is cutoff.
Wouldn't that be slower than just applying the cutoff every time (because now we are not only applying cutoff but also running an if statement)?

File: core/src/soundfont/mod.rs (hunk 4 above)
Comment (MyBlackMIDIScore) [reply core/src/soundfont/mod.rs:207]: Yes, I'll change this.

File: core/src/channel/voice_buffer.rs [hunk 6]
```diff
@@ -28,13 +28,17 @@ impl DerefMut for GroupVoice {
 pub struct VoiceBuffer {
     id_counter: usize,
     buffer: VecDeque<GroupVoice>,
+    damper: bool, // false = pedal up, true = pedal down
```
[the remark this answers is not in the document]
Comment (MyBlackMIDIScore) [reply core/src/channel/voice_buffer.rs:31]: I feel like it is not needed, as it is already pretty self explanatory. Plus it is easier to work with if statements.
I could rename the value from `damper` to `damper_pressed` or something so it's more descriptive.

[the remark this answers is not in the document]
Comment (arduano) [reply core/src/channel/voice_buffer.rs]: Yeah good call

**Review by arduano — APPROVED** [review]
